### DISCUSSION TITLE
Welcome to not-grocy Discussions!

### DISCUSSION THREAD

**Discussion by hackathi** [mistressofjellyfish/not-grocy#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  If you want, leave a short message to say hi in this thread.

**Comment by mohe2015** [comment]
Hi,

I think the following should be thought about:
1. Which version of JavaScript should be write: e.g. should we use fetch, Promises, async/await, querySelector, etc. As we could always transpile down and browser support is pretty good now I believe this may be useful.
2. Should we use TypeScript? What is your opinion on that? This wouldn't need to be done from the start of course but may be easier than adding afterwards if wanted at all. This adds a transpilation step / replaces one but should normally not be a problem. You should probably use incremental TypeScript compilation / another transpilation tool as it's quite slow otherwise.

**Reply by hackathi** [reply]
> Which version of JavaScript should be write: e.g. should we use fetch, Promises, async/await, querySelector, etc. As we could always transpile down and browser support is pretty good now I believe this may be useful.

Currently, the codebase uses the es2021/es12 `String.prototype.replaceAll` function and transpiles everything down to something the browserslist default understands (that is, if I configured babel correctly, if not, it's es2015). Since we're transpiling anyways, I don't see reason to limit ourselves to older versions. But – and that's a big but – I don't have an overview on what can be polyfilled / easily transpiled and what cannot and rely on babel complaining.

Personally, I want to nuke jQuery. DOM manipulation will be handled by vue at some point, and yes, all the API calls should move to fetch. I like promises and async await, but currently the project is not in a shape to sustain async functions. I also think that huge efforts to replace jQuery with native javascript aren't really beneficial and time is better spent just rewriting view after view to vue. I suspect that with a somewhat thought out separation into components this will become a relatively easy task once an initial set of components have been ported.

So, to summarize, my stance on this is: use es12 until babel complains it can't produce something supported in defaults sans IE11. 

> Should we use TypeScript? What is your opinion on that?

I think it would really benefit the project and reduce the maintenance burden, catch bugs before they even get committed, and make the suggestions in my IDE of choice better. Unfortunately, my experience with typescript is pretty much zero, and the legacy code is in no shape to tackle that. However, since this will get ported to vue anyways, that's a great opportunity to introduce that. But as you said, this requires incremental compilation.

Ideally, I would nuke gulp from the repo while I'm at it. I added it because it was an easy solution to the ~70 files in viewjs problem, mainly because I didn't want to force manual configuration for webpacks split chunks. But seeing as I wrapped them in functions now anyways, webpack is once again an option. Rollup came to be because I originally intended to keep much of the global functions (which I never got to work with webpack). the whole viewjs folder is ~450kb non-minified, and produces some hefty bundles eg. when bjip-js is included. Moving these to the main grocy bundle is an option now.
